### PR TITLE
Performance improvements to vectorized Span.Reverse

### PR DESCRIPTION
**Pull request by yesmey**
See original PR: #70944
Reverted: #78605

I reverted the overlapping for int and long and and chose to keep it in-between for byte/char where it still shows an improvement to overlap.

Machine:
``` ini
BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22621.819)
AMD Ryzen 7 3700X, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-AORABE : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-BUGVTU : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
```

### Byte 
```csharp
[HideColumns("Job", "Error", "StdDev", "Median", "RatioSD")]
public class ReverseBytes
{
  [Params(16, 24, 32, 36, 38, 40, 42, 44, 48, 56, 64)]
  public int NumberOfValues { get; set; }

  private byte[] _values;

  [GlobalSetup]
  public void Setup()
  {
      _values = ValuesGenerator.Array<byte>(NumberOfValues);
  }

  [Benchmark]
  public void Reverse() => _values.AsSpan().Reverse();
}
```

|  Method |                                                                                                          Toolchain | NumberOfValues |      Mean | Ratio |
|-------- |------------------------------------------------------------------------------------------------------------------- |--------------- |----------:|------:|
| **Reverse** | **main** |             **16** |  **7.880 ns** |  **1.00** |
| Reverse |   PR |             16 |  2.197 ns |  0.28 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **24** | **10.383 ns** |  **1.00** |
| Reverse |   PR |             24 |  2.772 ns |  0.27 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **32** |  **3.795 ns** |  **1.00** |
| Reverse |   PR |             32 |  2.852 ns |  0.75 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **36** |  **4.753 ns** |  **1.00** |
| Reverse |   PR |             36 |  3.373 ns |  0.71 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **38** |  **5.328 ns** |  **1.00** |
| Reverse |   PR |             38 |  3.985 ns |  0.75 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **40** |  **6.299 ns** |  **1.00** |
| Reverse |   PR |             40 |  3.387 ns |  0.54 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **42** |  **6.714 ns** |  **1.00** |
| Reverse |   PR |             42 |  4.236 ns |  0.63 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **44** |  **7.172 ns** |  **1.00** |
| Reverse |   PR |             44 |  3.479 ns |  0.49 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **48** |  **8.631 ns** |  **1.00** |
| Reverse |   PR |             48 |  4.737 ns |  0.55 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **56** | **10.799 ns** |  **1.00** |
| Reverse |   PR |             56 |  4.616 ns |  0.43 |
|         |                                                                                                                    |                |           |       |
| **Reverse** | **main** |             **64** |  **4.063 ns** |  **1.00** |
| Reverse |   PR |             64 |  3.028 ns |  0.75 |

### Char
```csharp
[HideColumns("Job", "Error", "StdDev", "Median", "RatioSD")]
public class ReverseChars
{
  [Params(8, 12, 16, 20, 24, 28, 32, 40, 48)]
  public int NumberOfValues { get; set; }

  private char[] _values;

  [GlobalSetup]
  public void Setup()
  {
      // taken from dotnet/performance
      _values = ValuesGenerator.Array<char>(NumberOfValues);
  }

  [Benchmark]
  public void Reverse() => _values.AsSpan().Reverse();
}
```

|  Method |                                                                                                          Toolchain | NumberOfValues |     Mean | Ratio |
|-------- |------------------------------------------------------------------------------------------------------------------- |--------------- |---------:|------:|
| **Reverse** | **main** |              **8** | **3.918 ns** |  **1.00** |
| Reverse |   PR |              8 | 4.145 ns |  1.06 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **12** | **5.724 ns** |  **1.00** |
| Reverse |   PR |             12 | 6.470 ns |  1.13 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **16** | **3.239 ns** |  **1.00** |
| Reverse |   PR |             16 | 2.786 ns |  0.86 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **20** | **4.361 ns** |  **1.00** |
| Reverse |   PR |             20 | 4.111 ns |  0.94 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **24** | **5.996 ns** |  **1.00** |
| Reverse |   PR |             24 | 4.637 ns |  0.77 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **28** | **8.948 ns** |  **1.00** |
| Reverse |   PR |             28 | 4.648 ns |  0.52 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **32** | **3.479 ns** |  **1.00** |
| Reverse |   PR |             32 | 2.518 ns |  0.72 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **40** | **5.494 ns** |  **1.00** |
| Reverse |   PR |             40 | 5.262 ns |  0.96 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **48** | **9.208 ns** |  **1.00** |
| Reverse |   PR |             48 | 3.624 ns |  0.39 |

### Int
|  Method |                                                                                                          Toolchain | NumberOfValues |     Mean | Ratio |
|-------- |------------------------------------------------------------------------------------------------------------------- |--------------- |---------:|------:|
| **Reverse** | **main** |              **4** | **2.692 ns** |  **1.00** |
| Reverse |   PR |              4 | 1.876 ns |  0.70 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |              **8** | **3.325 ns** |  **1.00** |
| Reverse |   PR |              8 | 2.131 ns |  0.64 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **12** | **4.020 ns** |  **1.00** |
| Reverse |   PR |             12 | 3.014 ns |  0.75 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **16** | **4.275 ns** |  **1.00** |
| Reverse |   PR |             16 | 3.018 ns |  0.71 |

### Long
|  Method |                                                                                                          Toolchain | NumberOfValues |     Mean | Ratio |
|-------- |------------------------------------------------------------------------------------------------------------------- |--------------- |---------:|------:|
| **Reverse** | **main** |              **4** | **2.393 ns** |  **1.00** |
| Reverse |   PR |              4 | 2.673 ns |  1.12 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |              **8** | **2.273 ns** |  **1.00** |
| Reverse |   PR |              8 | 2.240 ns |  0.98 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **12** | **3.605 ns** |  **1.00** |
| Reverse |   PR |             12 | 2.890 ns |  0.80 |
|         |                                                                                                                    |                |          |       |
| **Reverse** | **main** |             **16** | **2.718 ns** |  **1.00** |
| Reverse |   PR |             16 | 2.708 ns |  1.00 |

@adamsitnik can you confirm if you get similar numbers on your benchmark? I think earlier we saw some slight difference between our numbers last time

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 1]
```diff
@@ -1129,21 +1130,24 @@ private static unsafe nuint UnalignedCountVector128(ref byte searchSpace)
...
         public static void Reverse(ref byte buf, nuint length)
         {
-            if (Avx2.IsSupported && (nuint)Vector256<byte>.Count * 2 <= length)
+            Debug.Assert(length > 1);
+
+            nint remainder = (nint)length;
+            nint offset = 0;
+
+            // overlapping has a positive performance benefit around 48 elements
+            if (Avx2.IsSupported && remainder >= Vector256<byte>.Count * 1.5)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:1139]: This is going to do floating point math. That does not look cheap.

Comment (yesmey) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs]: I was a bit too fast on the keyboard, didn't check the assembly output. It only evaluates it as a constant if I use `(nint)(Vector256<byte>.Count * 1.5)`

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs]: > It only evaluates it as a constant if I use (nint)(Vector256<byte>.Count * 1.5)

just `Vector256<byte>.Count * 1.5` also evaluates into a constant, but a FP constant => datasection memory load

**Review by adamsitnik — APPROVED** [review]
The changes and the benchmark numbers LGTM, thank you @yesmey !

<details>

```ini
BenchmarkDotNet=v0.13.2.1950-nightly, OS=Windows 11 (10.0.22621.819)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK=8.0.100-alpha.1.22570.9
  [Host]     : .NET 8.0.0 (8.0.22.55902), X64 RyuJIT AVX2
  Job-EFMIQM : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-WVFVYW : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2

LaunchCount=3 MemoryRandomization=True
```


|        Type |  Job | Size |      Mean | Ratio |
|------------ |----- |----- |----------:|------:|
|  Span&lt;Byte&gt; |   PR |    4 |  3.525 ns |  0.90 |
|  Span&lt;Byte&gt; | main |    4 |  3.935 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |    4 |  4.050 ns |  0.96 |
|  Span&lt;Char&gt; | main |    4 |  4.220 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |    4 |  3.721 ns |  0.99 |
| Span&lt;Int32&gt; | main |    4 |  3.777 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |    4 |  3.468 ns |  0.88 |
| Span&lt;Int64&gt; | main |    4 |  3.936 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |    8 |  3.513 ns |  0.67 |
|  Span&lt;Byte&gt; | main |    8 |  5.237 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |    8 |  5.800 ns |  1.00 |
|  Span&lt;Char&gt; | main |    8 |  5.841 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |    8 |  3.458 ns |  0.94 |
| Span&lt;Int32&gt; | main |    8 |  3.669 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |    8 |  3.495 ns |  0.83 |
| Span&lt;Int64&gt; | main |    8 |  4.187 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   16 |  3.534 ns |  0.35 |
|  Span&lt;Byte&gt; | main |   16 | 10.259 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   16 |  4.002 ns |  0.83 |
|  Span&lt;Char&gt; | main |   16 |  4.832 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   16 |  3.527 ns |  0.92 |
| Span&lt;Int32&gt; | main |   16 |  3.852 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   16 |  4.217 ns |  0.82 |
| Span&lt;Int64&gt; | main |   16 |  5.128 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   24 |  3.996 ns |  0.34 |
|  Span&lt;Byte&gt; | main |   24 | 11.935 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   24 |  4.679 ns |  0.64 |
|  Span&lt;Char&gt; | main |   24 |  7.353 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   24 |  3.943 ns |  0.59 |
| Span&lt;Int32&gt; | main |   24 |  6.690 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   24 |  4.935 ns |  0.81 |
| Span&lt;Int64&gt; | main |   24 |  6.075 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   32 |  3.559 ns |  0.88 |
|  Span&lt;Byte&gt; | main |   32 |  4.044 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   32 |  4.278 ns |  0.83 |
|  Span&lt;Char&gt; | main |   32 |  5.152 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   32 |  3.971 ns |  0.84 |
| Span&lt;Int32&gt; | main |   32 |  4.755 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   32 |  5.630 ns |  0.83 |
| Span&lt;Int64&gt; | main |   32 |  6.799 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   36 |  4.103 ns |  0.80 |
|  Span&lt;Byte&gt; | main |   36 |  5.149 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   36 |  5.525 ns |  0.92 |
|  Span&lt;Char&gt; | main |   36 |  5.974 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   36 |  5.759 ns |  0.92 |
| Span&lt;Int32&gt; | main |   36 |  6.291 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   36 |  6.242 ns |  0.84 |
| Span&lt;Int64&gt; | main |   36 |  7.471 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   38 |  4.381 ns |  0.74 |
|  Span&lt;Byte&gt; | main |   38 |  5.911 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   38 |  6.316 ns |  0.96 |
|  Span&lt;Char&gt; | main |   38 |  6.575 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   38 |  6.303 ns |  0.93 |
| Span&lt;Int32&gt; | main |   38 |  6.781 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   38 |  6.236 ns |  0.81 |
| Span&lt;Int64&gt; | main |   38 |  7.689 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   40 |  4.153 ns |  0.63 |
|  Span&lt;Byte&gt; | main |   40 |  6.637 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   40 |  7.144 ns |  0.98 |
|  Span&lt;Char&gt; | main |   40 |  7.289 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   40 |  4.690 ns |  0.62 |
| Span&lt;Int32&gt; | main |   40 |  7.618 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   40 |  6.253 ns |  0.82 |
| Span&lt;Int64&gt; | main |   40 |  7.597 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   42 |  4.475 ns |  0.56 |
|  Span&lt;Byte&gt; | main |   42 |  8.054 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   42 |  9.112 ns |  1.12 |
|  Span&lt;Char&gt; | main |   42 |  8.159 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   42 |  5.279 ns |  0.59 |
| Span&lt;Int32&gt; | main |   42 |  8.975 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   42 |  6.724 ns |  0.88 |
| Span&lt;Int64&gt; | main |   42 |  7.674 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   44 |  4.190 ns |  0.44 |
|  Span&lt;Byte&gt; | main |   44 |  9.675 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   44 |  9.887 ns |  0.92 |
|  Span&lt;Char&gt; | main |   44 | 10.800 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   44 |  5.317 ns |  0.55 |
| Span&lt;Int32&gt; | main |   44 |  9.719 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   44 |  6.931 ns |  0.86 |
| Span&lt;Int64&gt; | main |   44 |  8.054 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   48 |  4.644 ns |  0.42 |
|  Span&lt;Byte&gt; | main |   48 | 11.215 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   48 |  5.343 ns |  0.42 |
|  Span&lt;Char&gt; | main |   48 | 12.655 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   48 |  4.915 ns |  0.85 |
| Span&lt;Int32&gt; | main |   48 |  5.781 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   48 |  7.562 ns |  0.90 |
| Span&lt;Int64&gt; | main |   48 |  8.376 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   56 |  4.616 ns |  0.34 |
|  Span&lt;Byte&gt; | main |   56 | 13.658 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   56 |  5.355 ns |  0.34 |
|  Span&lt;Char&gt; | main |   56 | 15.667 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   56 |  5.782 ns |  0.69 |
| Span&lt;Int32&gt; | main |   56 |  8.357 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   56 |  7.494 ns |  0.83 |
| Span&lt;Int64&gt; | main |   56 |  9.032 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |   64 |  4.108 ns |  0.92 |
|  Span&lt;Byte&gt; | main |   64 |  4.458 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |   64 |  5.358 ns |  0.88 |
|  Span&lt;Char&gt; | main |   64 |  6.077 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |   64 |  5.806 ns |  0.88 |
| Span&lt;Int32&gt; | main |   64 |  6.600 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |   64 |  7.711 ns |  0.79 |
| Span&lt;Int64&gt; | main |   64 |  9.736 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Byte&gt; |   PR |  512 |  9.985 ns |  0.90 |
|  Span&lt;Byte&gt; | main |  512 | 11.143 ns |  1.00 |
|             |      |      |           |       |
|  Span&lt;Char&gt; |   PR |  512 | 17.165 ns |  0.90 |
|  Span&lt;Char&gt; | main |  512 | 19.078 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int32&gt; |   PR |  512 | 29.440 ns |  0.99 |
| Span&lt;Int32&gt; | main |  512 | 29.709 ns |  1.00 |
|             |      |      |           |       |
| Span&lt;Int64&gt; |   PR |  512 | 62.140 ns |  1.00 |
| Span&lt;Int64&gt; | main |  512 | 62.204 ns |  1.00 |

</details>

**Pull request merged**